### PR TITLE
implement EnqueueExtensions interface in nodeaffinity

### PR DESCRIPTION
**Pull request by tanjing2020**
#### What type of PR is this?
/kind feature
/sig scheduling

#### Which issue(s) this PR fixes:

xref #94009

#### Special notes for your reviewer:
@Huang-Wei

#### Does this PR introduce a user-facing change?

```
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go [hunk 1]
```diff
@@ -72,6 +73,14 @@ func (s *preFilterState) Clone() framework.StateData {
 	return s
 }
 
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable.
+func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel | framework.UpdateNodeTaint},
```
[the remark this answers is not in the document]
Comment (chendave) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:80]: +1, I think `framework.UpdateNodeTaint` should go to `TaintToleration` plugin.

@Huang-Wei Here if "framework.Add" is registered, is that mean any kind of `Add` will leads to a move in the queue? or there is logic somewhere to make sure only the `NodeAdd` will be triggered.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go]: > if "framework.Add" is registered, 

framework.add should be associated with a particular resource, Node/PVC/PV, etc, to consists of a tuple. So NodeAdd is a shortened term of `{Resource: framework.Node, ActionType: framework.Add}`.

> is that mean any kind of `Add` will leads to a move in the queue?

We don't have a universe event matching **any** kind of resource.

> or there is logic somewhere to make sure only the `NodeAdd` will be triggered.

Yes, here it is:

https://github.com/kubernetes/kubernetes/blob/c2ff2910fd71c48db148e44d56ee18ef88503e89/pkg/scheduler/eventhandlers.go#L102

**Comment by tanjing2020** [comment]
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+
|          METRIC          |               GROUP                | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Average  | pods/s |  133.93 |  152.68 | +14.00% |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc50   | pods/s |   30.00 |   49.00 | +63.33% |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc90   | pods/s |  564.50 |  515.50 | -8.68%  |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc99   | pods/s |  564.50 |  515.50 | -8.68%  |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Average  | ms     |  458.46 |  334.55 | -27.03% |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc50   | ms     |   15.53 |   13.64 | -12.14% |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc90   | ms     | 1288.26 | 1004.83 | -22.00% |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc99   | ms     | 1749.03 | 1472.26 | -15.82% |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Average  | ms     | 3153.33 | 3339.24 | +5.90%  |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc50   | ms     | 3074.85 | 3346.54 | +8.84%  |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc90   | ms     | 4742.60 | 5580.41 | +17.67% |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc99   | ms     | 5783.46 | 6009.02 | +3.90%  |
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+

**Comment by tanjing2020** [comment]
```
root@ecs-linux-0001-f8d9:/usr1/k8s/src/k8s.io/k8s-sched-perf-stat# go run *.go /tmp_old/old.txt /tmp_new/new.txt
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+
|          METRIC          |               GROUP                | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Average  | pods/s |  351.81 |  358.33 | +1.85%  |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc50   | pods/s |  314.67 |  318.90 | +1.35%  |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc90   | pods/s |  584.89 |  555.80 | -4.97%  |
| SchedulingThroughput     | Unschedulable/500Nodes/200InitPods | Perc99   | pods/s |  584.89 |  555.80 | -4.97%  |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Average  | ms     |   13.86 |   21.69 | +56.44% |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc50   | ms     |    6.24 |    6.07 | -2.64%  |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc90   | ms     |   14.08 |   15.24 | +8.22%  |
| scheduler_e2e_scheduling | Unschedulable/500Nodes/200InitPods | Perc99   | ms     |   17.71 |   21.39 | +20.73% |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Average  | ms     | 1788.82 | 1893.09 | +5.83%  |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc50   | ms     | 1874.09 | 2039.84 | +8.84%  |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc90   | ms     | 2886.48 | 3077.72 | +6.63%  |
| scheduler_pod_scheduling | Unschedulable/500Nodes/200InitPods | Perc99   | ms     | 3071.59 | 3207.61 | +4.43%  |
+--------------------------+------------------------------------+----------+--------+---------+---------+---------+
```

**Comment by Huang-Wei** [comment]
@tanjing2020 Testcase `Unschedulable` cannot quite prove this PR improve the perf or not :)

Anyway, I think the logic is straightforward, and we don't rush for 1.21. So I'd not insist on enforcing a desired benchmark test result. But as a manner to understand more scheduler internals, it's always welcome to build a test and verify the result, sort of like a bar raiser.

**Pull request merged**